hep/proposals/0001-decorators.md

```diff
@@ -0,0 +1,313 @@
+
+# Proposal
+
+This HEP proposes that Argo's DAG, Steps and Container templates be mapped into Hera as decorated functions with HeraIO classes as the inputs and outputs.
+
+# Code Examples
+
+## Basic Workflow
+
+```python
+from hera.workflows import WorkflowTemplate
+import hera.workflows.io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# Users must subclass `hio.Input` to define the template's inputs
+class MyInput(hio.Input):
+    user: str
+
+@wt.entrypoint  # Sets hello_world as the default entrypoint for the workflow template, it is an error to set entrypoint on multiple functions
+@wt.script  # Adds a new script template to the workflow template called hello_world
+def hello_world(my_input: MyInput) -> hio.Output:  # A subclass of hio.Output must be used for the output of the function
+    output = hio.Output()
+    output.result = f"Hello Hera User: {my_input.user}!"
+    return output
+
+# `run` is a new function for users to instantiate a workflow template as a workflow
+workflow = wt.run(MyInput(name="happy-hera-user"), wait=True)
+
+# Returned workflow is a normal hera Workflow class and can be accessed as normal
+assert workflow.status.phase == "Succeeded"
+assert workflow.status.outputs.result == "Hello Hera User: happy-hera-user!"
+```
+
+## DAG Template
+
+```python
+from hera.workflows import WorkflowTemplate
+import hera.workflows.io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+@wt.script  # Adds a new script template to the workflow template
+def setup() -> hio.Output:
+    return hio.Output(result="Setting things up")
+
+class ConcatInput(hio.Input):
+    word_a: str
+    word_b: str
+
+@wt.script  # Adds a new script template to the workflow template
+def concat(concat_input: ConcatInput) -> hio.Output:
+    return hio.Output(result=f"{concat_input.word_a} {concat_input.word_b}")
+
+class WorkerInput(hio.Input):
+    value_a: str
+    value_b: str
+
+class WorkerOutput(hio.Output):
+    value: str
+
+
+# The great outcome of this new syntax is is the separation of concerns so that
+# the template definition vs the template call becomes very explicit.
+# When a template is defined, it is defined via the decorator which registers it
+# When a template is invoked, it is invoked either via the callable syntax or via
+# wt.run() if it is the entrypoint
+# Another great side-effect is that all the templates can run locally as functions
+@wt.entrypoint
```

Comment:
Is there a design principle that clearly delineates what will be available as a decorator and what won't?

Reply:
Currently every template type and certain properties associated with the template would be decorators. I believe entrypoint is the only one aside from various template types currently.

Reply:
To explain more of the thinking behind the proposal, I'm not sure there's a particular design principle we're using (maybe DRY or dependency inversion?) but the intention is to avoid multiples places where we're performing definition logic, while keeping in mind a user-friendly and well-scoped way of writing WorkflowTemplates:

1. Declare WorkflowTemplate
2. Add templates at definition via decorator

As opposed to:
1. Write "free" (as in unbound) template definitions
2. Declare WorkflowTemplate
3. Add the templates

Which means anything we want to declare as part of the WorkflowTemplate needs to be from that local scope and not "forward declaring," so for entrypoint we want to avoid:
```py
wt = WorkflowTemplate(..., entrypoint="my-template")
```

As Sam said, seems like entrypoint is the only attribute that refers to something "yet to be created."